cheroot/server.py

```diff
@@ -131,53 +131,71 @@ def bton(b, encoding='ISO-8859-1'):
+        Read headers from the given stream into the given header dict.
 
-    Headers which are repeated are folded together using a comma if their
-    specification so dictates.
+        If hdict is None, a new header dict is created. Returns the populated
+        header dict.
 
-    This function raises ValueError when the read bytes violate the HTTP spec.
-    You should probably return "400 Bad Request" if this happens.
-    """
-    if hdict is None:
-        hdict = {}
+        Headers which are repeated are folded together using a comma if their
+        specification so dictates.
 
-    while True:
-        line = rfile.readline()
-        if not line:
-            # No more data--illegal end of headers
-            raise ValueError('Illegal end of headers.')
-
-        if line == CRLF:
-            # Normal end of headers
-            break
-        if not line.endswith(CRLF):
-            raise ValueError('HTTP requires CRLF terminators')
-
-        if line[0] in (SPACE, TAB):
-            # It's a continuation line.
-            v = line.strip()
-        else:
-            try:
-                k, v = line.split(COLON, 1)
-            except ValueError:
-                raise ValueError('Illegal header line.')
-            # TODO: what about TE and WWW-Authenticate?
-            k = k.strip().title()
-            v = v.strip()
-            hname = k
+        This function raises ValueError when the read bytes violate the HTTP spec.
+        You should probably return "400 Bad Request" if this happens.
+        """
+        if hdict is None:
+            hdict = {}
+
+        while True:
+            line = rfile.readline()
+            if not line:
+                # No more data--illegal end of headers
+                raise ValueError('Illegal end of headers.')
+
+            if line == CRLF:
+                # Normal end of headers
+                break
+            if not line.endswith(CRLF):
+                raise ValueError('HTTP requires CRLF terminators')
+
+            if line[0] in (SPACE, TAB):
+                # It's a continuation line.
+                v = line.strip()
+            else:
+                try:
+                    k, v = line.split(COLON, 1)
+                except ValueError:
+                    raise ValueError('Illegal header line.')
+                v = v.strip()
+                k = self._transform_key(k)
+                hname = k
+
+            if not self._allow_header(k):
+                continue
 
-        if k in comma_separated_headers:
-            existing = hdict.get(hname)
-            if existing:
-                v = b', '.join((existing, v))
-        hdict[hname] = v
+            if k in comma_separated_headers:
+                existing = hdict.get(hname)
+                if existing:
+                    v = b', '.join((existing, v))
+            hdict[hname] = v
 
-    return hdict
+        return hdict
+
+    def _allow_header(self, key_name):
+        return key_name
```

Comment:
Hmm. Perhaps. Unless we want to filter out empty headers. I guess the previous implementation allowed for empty headers, so this refactor should as well.